schema/migration.go

```diff
@@ -37,57 +46,365 @@ func (x Versions) Latest() Version { return x[len(x)-1] }
+			copy(cp, kp)
+			fields = append(fields, &VersionDeltaField{KeyPath: cp, From: v1.DataType, To: UnknownType})
+		case v1.DataType != v2.DataType: // type changed
+			cp := make([]string, len(kp))
+			copy(cp, kp)
+			fields = append(fields, &VersionDeltaField{KeyPath: cp, From: v1.DataType, To: v2.DataType})
+		case v1.DataType == ObjectType:
+			fields = append(fields, buildSchemaDeltaLow(kp, v1.Fields, v2.Fields)...)
+		case v1.DataType == ArrayType:
+			fields = append(fields, buildSchemaDeltaLow(kp, v1.Fields, v2.Fields)...)
+		}
+	}
+
+	return fields
+}
+
+// buildSchemaDelta build delta between two consecutive schema versions.
+func buildSchemaDelta(schema1 Version, schema2 Version) (*VersionDelta, error) {
+	var (
+		first, second *Factory
+		delta         VersionDelta
+		err           error
+	)
+
+	if first, err = Build("", schema1.Schema); err != nil {
+		return nil, err
+	}
+
+	if second, err = Build("", schema2.Schema); err != nil {
+		return nil, err
+	}
+
+	delta.Version = schema2.Version
+	delta.Fields = buildSchemaDeltaLow(nil, first.Fields, second.Fields)
+
+	return &delta, nil
 }
 
 // buildSchemaDeltas builds all the schema deltas from the collection schema changes history.
-func buildSchemaDeltas(_ Versions) []VersionDelta {
-	// TODO: Implement
-	return []VersionDelta{{0, nil}}
+func buildSchemaDeltas(schemas Versions) ([]VersionDelta, error) {
+	var deltas []VersionDelta
+
+	for i := 1; i < len(schemas); i++ {
+		delta, err := buildSchemaDelta(schemas[i-1], schemas[i])
+		if err != nil {
+			return nil, err
+		}
+
+		// add only if there are changes
+		if len(delta.Fields) > 0 {
+			deltas = append(deltas, *delta)
+		}
+	}
+
+	if len(deltas) != 0 {
+		return deltas, nil
+	}
+
+	// Just one schema Version or no incompatible schema changes
+	return []VersionDelta{{0, nil}}, nil
 }
 
-// CompatibleSchemaSince determines if there was incompatible schema change since given version.
+// CompatibleSchemaSince determines if there was incompatible schema change since given Version.
 func (d *DefaultCollection) CompatibleSchemaSince(version int32) bool {
 	// first element in the array contains last incompatible schema change
-	return version >= d.SchemaDeltas[0].version
+	return int(version) >= d.SchemaDeltas[len(d.SchemaDeltas)-1].Version
 }
 
-// UpdateRowSchema fixes the schema of  provided, unmarshalled, document from given version to the latest
-// collection schema.
-func (d *DefaultCollection) UpdateRowSchema(_ map[string]any, version int32) error {
-	if d.CompatibleSchemaSince(version) {
-		return nil
+func convertFromBool(toType FieldType, val bool) any {
+	switch toType {
+	case BoolType:
+		return val
+	case Int32Type, Int64Type, DoubleType:
+		if val {
+			return 1
+		}
+		return 0
+	case StringType:
+		if val {
+			return "true"
+		}
+		return "false"
+	case ByteType:
+		if val {
+			return []byte{0x1}
+		}
+		return []byte{0x0}
+	case UUIDType, DateTimeType, ArrayType, ObjectType:
+		// non convertible
```

Comment:
An alternative, ee could return true if the field is non null otherwise false. We would just have to document our behaviour